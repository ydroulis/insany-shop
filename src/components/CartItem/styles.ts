"use client"
import styled from 'styled-components'
import Image from 'next/image'

export const Container = styled.article`
    display: flex;
    gap: 10px;

    width: 100%;
    height: 200px;

    background-color: #fff;
    border-radius: 8px;

    box-sizing: border-box;

    padding-right: 1rem;

    @media (min-width: 768px) {
        width: 406px;
        gap: 16px;
        padding-right: 1rem;
    }

    @media (min-width: 1024px) {
        width: 636px;
        gap: 31px;

        height: 211px;
    }

    @media (min-width: 1440px) {
        width: 736px;
        gap: 31px;

        height: 211px;
    }
`

export const ProductImage = styled(Image)`
    width: 100px;
    min-width: 100px;
    height: 100px;
    border-radius: 8px 0px 0px 8px;

    @media (min-width: 768px) {
        min-width: 100px;
        height: 100px;
    }

    @media (min-width: 1024px) {
        min-width: 256px;
        height: 211px;
    }

    @media (min-width: 1440px) {
        min-width: 256px;
        height: 211px;
    }
`

export const ProductInfo = styled.div`
    width: 100%;
    position: relative;

    padding: .5rem 0 .5rem 0;

    color: #41414D;

    @media (min-width: 768px) {
        padding: 1rem 0 1.5rem 0;
    }
`

export const Remove = styled.button`
    background-color: transparent;
    border: none;
    cursor: pointer;

    position: absolute;
    

    &:hover {
        opacity: 0.8;
    }

    @media (max-width: 767px) {
        bottom: 30px;
        right: 0px;
    }

    @media (min-width: 768px) {
        top: 16px;
        right: 16px;
    }
`

export const ProductName = styled.h2`
    font-size: 1.25rem;
    font-weight: 300;

    margin-bottom: 0rem;

    @media (min-width: 768px) {
        margin-bottom: 0.75rem;
    }
`

export const Description = styled.p`
    font-size: 0.75rem;

    margin-bottom: 0.75rem;
`

export const Values = styled.div`
    max-width: 433px;

    display: flex;
    align-items: center;

    position: absolute;
    justify-content: space-evenly;

    bottom: 24px;

    @media (max-width: 767px) {
        right: 28px;
        flex-direction: row;
        gap: 20px;
        right: 60px;
    }

    @media (min-width: 768px) {
        width: 100%;
        justify-content: space-between;
    }

    @media (min-width: 1024) {
        width: 100%;
        justify-content: space-between;
    }

    @media (min-width: 1440px) {
        width: 100%;
        justify-content: space-between;
    }
`

export const Price = styled.p`
    color: #09090A;
    font-size: 1rem;
    font-weight: 600;
`