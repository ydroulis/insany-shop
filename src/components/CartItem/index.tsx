import React, { useEffect, useState } from 'react';
import { BsTrash3 } from "react-icons/bs";
import * as S from './styles';
import QuantitySelector from '../QuantitySelector';
import { useCartStore } from '../../providers/cartStoreProvider';
import CartFeedback from '../CartFeedback';

interface CartItemProps {
    name: string;
    description: string;
    price: number;
    image: string;
    id: number;
    stock: number
}

const CartItem: React.FC<CartItemProps> = ({ name, description, price, image, id, stock }) => {
    const { removeProductFromCart, changeItems, showFeedback } = useCartStore((state) => state);
    const [value, setValue] = useState('1');
    const [finalPrice, setFinalPrice] = useState(price);

    const options = Array.from({ length: Math.min(stock, 10) }, (_, i) => ({
        value: String(i + 1),
        label: String(i + 1),
    }));

    useEffect(() => {
        const amount = price * Number(value);
        setFinalPrice(amount);
        changeItems(id, Number(value));

    }, [changeItems, id, price, value]);

    const formattedPrice = new Intl.NumberFormat('pt-BR', {
        style: 'currency',
        currency: 'BRL',
    }).format(finalPrice);

    const handleRemoveProductFromCart = () => {
        removeProductFromCart(id, price);

        showFeedback(true);
        setTimeout(() => {
            showFeedback(false);
        }, 3000);
    }

    return (
        <S.Container data-testid="cart-item" aria-label={`Produto: ${name}`}>
            <S.ProductImage
                src={image}
                alt={`Imagem do produto ${name}`}
                width={256}
                height={211}
                loading='lazy'
                placeholder='blur'
                blurDataURL='...'
            />
            <S.ProductInfo>
                <S.Remove type="button" aria-label={`Remover ${name} do carrinho`} onClick={handleRemoveProductFromCart}>
                    <BsTrash3 color='#DE3838' size={24} aria-hidden="true" />
                </S.Remove>
                <S.ProductName>{name}</S.ProductName>
                <S.Description>{description}</S.Description>
                <S.Values>
                    <QuantitySelector
                        value={value}
                        setValue={setValue}
                        options={options}
                        label={`Quantidade de ${name}`}
                        id="quantity"
                    />
                    <S.Price aria-label={`PreÃ§o do produto ${formattedPrice}`}>
                        {formattedPrice}
                    </S.Price>
                </S.Values>
            </S.ProductInfo>
        </S.Container>
    );
}

export default CartItem;