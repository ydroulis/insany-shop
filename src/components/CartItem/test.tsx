import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import CartItem from './';

jest.mock('../QuantitySelector', () => {
    const QuantitySelectorMock = ({
        value,
        setValue,
        label,
        id,
    }: {
        value: string;
        setValue: React.Dispatch<React.SetStateAction<string>>;
        label: string;
        id: string;
    }) => (
        <div>
            <label htmlFor={id}>{label}</label>
            <select
                data-testid="quantity-selector"
                value={value}
                onChange={(e) => setValue(e.target.value)}
            >
                <option value="1">1</option>
                <option value="2">2</option>
                <option value="3">3</option>
                <option value="4">4</option>
                <option value="5">5</option>
            </select>
        </div>
    );
    QuantitySelectorMock.displayName = 'QuantitySelectorMock';
    return QuantitySelectorMock;
});

type MockCartProduct = {
    id: number;
    name: string;
    items: number;
};

type MockCart = {
    content: MockCartProduct[];
};

const mockCartStore = {
    removeProductFromCart: jest.fn(),
    changeItems: jest.fn(),
    showFeedback: jest.fn(),
    cart: { content: [] } as MockCart, // ðŸ‘ˆ garantindo que cart.content existe
};

jest.mock('../../providers/cartStoreProvider', () => ({
    useCartStore: jest.fn(() => mockCartStore),
}));

describe('<CartItem />', () => {
    const defaultProps = {
        name: 'Produto Teste',
        description: 'DescriÃ§Ã£o do produto teste',
        price: 123.45,
        image: 'https://via.placeholder.com/256x211.png',
        id: 1,
        stock: 5,
    };

    beforeEach(() => {
        jest.clearAllMocks();
        mockCartStore.cart.content = []; // limpa carrinho antes de cada teste
    });

    it('should render properly all elements', () => {
        render(<CartItem {...defaultProps} />);

        const container = screen.getByTestId('cart-item');
        expect(container).toBeInTheDocument();

        const image = screen.getByRole('img', { name: /imagem do produto produto teste/i });
        expect(image).toBeInTheDocument();

        expect(screen.getByText(defaultProps.name)).toBeInTheDocument();
        expect(screen.getByText(defaultProps.description)).toBeInTheDocument();

        expect(
            screen.getByText((content) => content.replace(/\s/g, '') === 'R$123,45')
        ).toBeInTheDocument();

        const select = screen.getByTestId('quantity-selector') as HTMLSelectElement;
        expect(select).toBeInTheDocument();
        expect(select.value).toBe('1');

        const removeButton = screen.getByRole('button', { name: /remover produto teste do carrinho/i });
        expect(removeButton).toBeInTheDocument();
        const svg = removeButton.querySelector('svg');
        expect(svg).toBeInTheDocument();
    });

    it('should allow change quantity and call changeItems', () => {
        render(<CartItem {...defaultProps} />);
        const select = screen.getByTestId('quantity-selector') as HTMLSelectElement;

        fireEvent.change(select, { target: { value: '3' } });
        expect(select.value).toBe('3');
        expect(mockCartStore.changeItems).toHaveBeenCalledWith(defaultProps.id, 3);
    });

    it('should call removeProductFromCart and showFeedback on remove button click', () => {
        render(<CartItem {...defaultProps} />);
        const removeButton = screen.getByRole('button', { name: /remover produto teste do carrinho/i });

        fireEvent.click(removeButton);
        expect(mockCartStore.removeProductFromCart).toHaveBeenCalledWith(defaultProps.id, defaultProps.price);
        expect(mockCartStore.showFeedback).toHaveBeenCalledWith(true);
    });

    it('should match snapshot', () => {
        const { container } = render(<CartItem {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });
});
