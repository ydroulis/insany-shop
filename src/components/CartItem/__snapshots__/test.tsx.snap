// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<CartItem /> should match snapshot 1`] = `
<div>
  <article
    aria-label="Produto: Produto Teste"
    class="sc-bRKDuR hLvVdg"
    data-testid="cart-item"
  >
    <img
      alt="Imagem do produto Produto Teste"
      class="sc-hvigdm flWskD"
      data-nimg="1"
      decoding="async"
      height="211"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F256x211.png&w=640&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F256x211.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F256x211.png&w=640&q=75 2x"
      style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      width="256"
    />
    <div
      class="sc-fhHczv dtjZdJ"
    >
      <button
        aria-label="Remover Produto Teste do carrinho"
        class="sc-ggWZvA FHRrb"
        type="button"
      >
        <svg
          aria-hidden="true"
          color="#DE3838"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(222, 56, 56);"
          viewBox="0 0 16 16"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.5 1h3a.5.5 0 0 1 .5.5v1H6v-1a.5.5 0 0 1 .5-.5M11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3A1.5 1.5 0 0 0 5 1.5v1H1.5a.5.5 0 0 0 0 1h.538l.853 10.66A2 2 0 0 0 4.885 16h6.23a2 2 0 0 0 1.994-1.84l.853-10.66h.538a.5.5 0 0 0 0-1zm1.958 1-.846 10.58a1 1 0 0 1-.997.92h-6.23a1 1 0 0 1-.997-.92L3.042 3.5zm-7.487 1a.5.5 0 0 1 .528.47l.5 8.5a.5.5 0 0 1-.998.06L5 5.03a.5.5 0 0 1 .47-.53Zm5.058 0a.5.5 0 0 1 .47.53l-.5 8.5a.5.5 0 1 1-.998-.06l.5-8.5a.5.5 0 0 1 .528-.47M8 4.5a.5.5 0 0 1 .5.5v8.5a.5.5 0 0 1-1 0V5a.5.5 0 0 1 .5-.5"
          />
        </svg>
      </button>
      <h2
        class="sc-dTvVRJ grTNHw"
      >
        Produto Teste
      </h2>
      <p
        class="sc-jwTyAe gDnhfW"
      >
        Descrição do produto teste
      </p>
      <div
        class="sc-hjsuWn kGyVvi"
      >
        <div>
          <label
            for="quantity"
          >
            Quantidade de Produto Teste
          </label>
          <select
            data-testid="quantity-selector"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
          </select>
        </div>
        <p
          aria-label="Preço do produto R$ 123,45"
          class="sc-jJLAfE kSSYMx"
        >
          R$ 123,45
        </p>
      </div>
    </div>
  </article>
</div>
`;
